### DISCUSSION TITLE
[v4] How to get array of ids of a relation? Is this impossible?

### DISCUSSION THREAD

**Discussion by exenestecnico** [adonisjs/core#2406] [Help]
I have two many-to-many related models: Client, House; and want to get an array of ids of the houses that belongs to a client.
The relation works and I get the related objects using `await client.houses().fetch()`, but dont need the whole objects just the ids.

Tried using `await client.houses().ids()` and `await client.houses().pluck('id')` but then I get all the ids of the table not just the ones related to the client. Shouldn't those methods return just the ids of the related objects? Tried looking at the source of Lucid but got lost.

My current solution that I'm trying to improove is:
```
const houses = await client.houses().fetch()
const ids = houses.rows.map(({ id }) => id)
```
There is no reason to get all the fields from the DB when the query could just select the id column.

This are the model declarations. There is a pivot table for the relation.
```
class Client extends Model {
  houses() {
    return this.belongsToMany('App/Models/House')
  }
}

class House extends Model {
  clients() {
    return this.belongsToMany('App/Models/Client')
  }
}
```

**Comment by ammezie** [comment]
Something like this should work:

```
  const houses = await client.houses().ids()
```

**Reply by exenestecnico** [reply]
Already tried that. Is stated in the question.

**Comment by artem-a** [comment]
Hi @exenestecnico you have to use  the `.ids()` method, it returns related data

Here is an example

I have three tables `users`, `cars` & `user_cars`

Here is the data 
```bash
sqlite> SELECT * from users;
1|user_1|user_1@mail.com|2021-04-11 23:31:24|2021-04-11 23:31:24
2|user_2|user_2@mail.com|2021-04-11 23:31:24|2021-04-11 23:31:24

sqlite> SELECT * from cars;
1|car_1|model_1|2021-04-11 23:38:11|2021-04-11 23:38:11
2|car_2|model_2|2021-04-11 23:38:11|2021-04-11 23:38:11
3|car_3|model_3|2021-04-11 23:38:11|2021-04-11 23:38:11

sqlite> SELECT * from user_cars;
1|1|1
2|1|3
3|2|2

```
and here is getting the user1 & user2 car ids

```javascript
    const user1 = await User.find(1)
    const user2 = await User.find(2)

    const u1CarIds = await user1.cars().ids()
    const u2CarIds = await user2.cars().ids()

    console.log(u1CarIds, u2CarIds)
```

here is the output
```bash
[ 1, 3 ] [ 2 ]
```


I enabled debug for the database provider so you can see SQL queries for each user

```bash
# for the first user
{
  method: 'pluck',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 1 ],
  __knexQueryUid: 'FW4ESbxECs8MxlVeYDUxW',
  sql: 'select `cars`.`id` from `cars` inner join `user_cars` on `cars`.`id` = `user_cars`.`car_id` where `user_cars`.`user_id` = ?',
  pluck: 'id'
}

# for the second user
{
  method: 'pluck',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 2 ],
  __knexQueryUid: 'Y7bOJev0cZnuRFCUsX4RU',
  sql: 'select `cars`.`id` from `cars` inner join `user_cars` on `cars`.`id` = `user_cars`.`car_id` where `user_cars`.`user_id` = ?',
  pluck: 'id'
}
```
as you can see for the first query the user id is 1 **bindings: [ 1 ]** and for the second query the user id is 2 **bindings: [ 2 ]**
that's why you couldn't get all the ids of table.

**Reply by artem-a** [reply]
And here are the model definitions

```javascript
class Car extends Model {
  users () {
    return this.belongsToMany('App/Models/User')
    .pivotTable('user_cars')
  }
}

class User extends Model {
  cars () {
    return this.belongsToMany('App/Models/Car')
    .pivotTable('user_cars')
  }
}
```

**Reply by exenestecnico** [reply]
It doesn't work for me, I don't know why.

When I do 
```
  const client = await Client.find(1)
  const ids = await client.houses().ids()
```
it returns the ids of **all** the rows in the table. This example should return only one id of the house currently related to client 1.

Also tried `client.houses().ids().toSQL()` to look at the generated query, it returned:
```
{ method: 'pluck',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 1 ],
  __knexQueryUid: '2b1e17a9-caa8-490d-a762-7228f9cbdedf',
  sql:
   'select
      `houses`.`id`, `houses`.*,
      `client_house`.`house_id` as `pivot_house_id`,
      `client_house`.`client_id` as `pivot_client_id`
    from `houses` inner join `client_house` on `houses`.`id` = `client_house`.`house_id`
    where `client_house`.`client_id` = ?',
  pluck: 'id' }
```
I added here some newlines in the sql part for readability.

When I run that query in a DB manager it returns the 1 correct row, although with extra fields that I dont want because of the _houses.*_ part.

Im using Lucid 5.0.3, maybe that's the issue. What version are you using?

**Reply by artem-a** [reply]
I've used "@adonisjs/lucid": "^6.3.0", you need to update your version and check.
